### DISCUSSION TITLE
What would be the most efficient way of using the BYOB API for WebTransport stream readers?

### DISCUSSION THREAD

**Discussion by MOZGIII** [wasm-bindgen/wasm-bindgen#3672] [Q&A]
As I understand, when working with WebTransport from WASM, the most efficient mode of operation would be to use a BYOB reader mode backed by a buffer that can be used from both within WASM and outside of WASM without transferring (i.e. copying) the backed memory across.

So, it seems like allocating an `&mut [u8]` and passing that as a buffer in WASM would work, but the browsers react to that poorly and take ownership of the whole wasm memory (and mess it up). This has already been discussed [here](https://github.com/rustwasm/wasm-bindgen/discussions/3579#discussioncomment-6846815). It is fixed in the spec, but the browser implementations for the update is not ready yet.

So. What would be the most efficient way of using BYOB today, and once browsers finally implement the spec? Can someone come up with a code sample?

I currently simply do this: https://github.com/MOZGIII/xwebtransport/blob/1afc3c4fcb5c42e6546129d273b0b0c1b08b9f70/crates/xwebtransport-web-sys/src/lib.rs#L217-L231

```rust
    async fn read(&mut self, buf: &mut [u8]) -> Result<Option<usize>, Self::Error> {
        let internal_buf = js_sys::Uint8Array::new_with_length(buf.len().try_into().unwrap());
        let maybe_output_buf =
            web_sys_stream_utils::read_byob(&self.reader.inner, internal_buf).await?; // calls `read_with_array_buffer`
        let Some(output_buf) = maybe_output_buf else {
            return Ok(None);
        };

        // Unwrap is safe assuming the `usize` is `u32` in wasm.
        let len = output_buf.byte_length().try_into().unwrap();

        output_buf.copy_to(&mut buf[..len]);

        Ok(Some(len))
    }
```

This, however, will allocate two buffers instead of one: one has to exist at the caller code side and passed as `buf: &mut [u8]` to this fn, and another one would be allocated "inside" of the `js_sys::Uint8Array::new_with_length` call. The backed memory of these buffers would be separate, and thus we'd need copy the data over (with `output_buf.copy_to`).

Is it possible to get rid of the extra copy (i.e. the `output_buf.copy_to`) and make the underlying browser native code to directly write to the `buf: &mut [u8]`?

**Comment by daxpedda** [comment]
> It is fixed in the spec, but the browser implementations for the update is not ready yet.

This was a mistake on my part, this wasn't fixed in the spec, this is intended and browsers have implemented it correctly so far. So right now it looks like this isn't going to change.

See https://github.com/rustwasm/wasm-bindgen/discussions/3579#discussioncomment-6851838.

> Is it possible to get rid of the extra copy (i.e. the `output_buf.copy_to`) and make the underlying browser native code to directly write to the `buf: &mut [u8]`?

As far as I'm aware this is currently not possible. See https://github.com/WebAssembly/design/issues/1162, I didn't read through it, so maybe you can find a gem in there. Also see https://github.com/whatwg/streams/issues/757, the underlying JS issue.

**Reply by MOZGIII** [reply]
Great references, thanks! I'll check them out.
